service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/flight/RetrieveControlledBigQueryDatasetLifetimeStep.java

```diff
@@ -0,0 +1,134 @@
+        context
+            .getWorkingMap()
+            .get(CONTROLLED_BIG_QUERY_DATASETS_WITHOUT_LIFETIME, new TypeReference<>() {});
+    Map<UUID, String> resourceIdToDefaultTableLifetimeMap = new HashMap<>();
+    Map<UUID, String> resourceIdToDefaultPartitionLifetimeMap = new HashMap<>();
+    Map<UUID, String> resourceIdToWorkspaceIdMap = new HashMap<>();
+
+    assert controlledBigQueryDatasets != null;
+    for (var resource : controlledBigQueryDatasets) {
+      logger.info(
+          "Getting default table lifetime and partition life for resource (BigQuery dataset) {} in workspace {}",
+          resource.getResourceId(),
+          resource.getWorkspaceId());
+      populateMapsWithResourceIdKey(
+          resourceIdToDefaultTableLifetimeMap,
+          resourceIdToDefaultPartitionLifetimeMap,
+          resourceIdToWorkspaceIdMap,
+          resource,
+          getBqDatasetDefaultTableLifetime(resource),
+          getBqDatasetDefaultPartitionLifetime(resource));
+    }
+    context
+        .getWorkingMap()
+        .put(
+            CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_TABLE_LIFETIME_MAP,
+            resourceIdToDefaultTableLifetimeMap);
+    context
+        .getWorkingMap()
+        .put(
+            CONTROLLED_BIG_QUERY_DATASET_RESOURCE_ID_TO_PARTITION_LIFETIME_MAP,
+            resourceIdToDefaultPartitionLifetimeMap);
+    context
+        .getWorkingMap()
+        .put(CONTROLLED_RESOURCE_ID_TO_WORKSPACE_ID_MAP, resourceIdToWorkspaceIdMap);
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void populateMapsWithResourceIdKey(
+      Map<UUID, String> resourceIdToDefaultTableLifetimeMap,
+      Map<UUID, String> resourceIdToDefaultPartitionLifetimeMap,
+      Map<UUID, String> resourceIdToWorkspaceIdMap,
+      ControlledResource resource,
+      @Nullable Long defaultTableLifetime,
+      @Nullable Long defaultPartitionLifetime) {
+    UUID resourceId = resource.getResourceId();
+    resourceIdToWorkspaceIdMap.put(resourceId, resource.getWorkspaceId().toString());
+    if (defaultTableLifetime != null) {
+      resourceIdToDefaultTableLifetimeMap.put(resourceId, defaultTableLifetime.toString());
+    }
+    if (defaultPartitionLifetime != null) {
+      resourceIdToDefaultPartitionLifetimeMap.put(resourceId, defaultPartitionLifetime.toString());
+    }
+  }
+
+  @Nullable
+  private Long getBqDatasetDefaultTableLifetime(ControlledBigQueryDatasetResource resource) {
+    Dataset dataset = getBqDataset(resource);
+    if (dataset != null) {
+      return dataset.getDefaultTableExpirationMs() / 1000;
+    }
+    return null;
+  }
+
+  @Nullable
+  private Long getBqDatasetDefaultPartitionLifetime(ControlledBigQueryDatasetResource resource) {
+    Dataset dataset = getBqDataset(resource);
+    if (dataset != null) {
+      return dataset.getDefaultPartitionExpirationMs() / 1000;
+    }
+    return null;
+  }
+
+  @Nullable
+  private Dataset getBqDataset(ControlledBigQueryDatasetResource resource) {
```

Comment:
I'd like to understand the failure/retry model here. If there is a cloud glitch and we fail to get the dataset, then we will not back fill it. I guess we will rerun this every time we restart a pod, so in theory it will converge.

But how will we know it is safe to remove this code? Can we have this code run the "what needs backfill?" query at the end and report when it is done?

Reply:
so I realize as I'm doing the region backfill, some bucket or dataset have a lifetime so they are deleted in the gcp project but we still have them in the database. 

i manually checked them for the preprod env and manually deleted the db row when i don't see it in the cloud project. Prod env is very clean.

there are so many in devel that I didn't bother to check manually (they can be from test glitches or just admin users manually deleting them from the gcp project too).

that's being said, we can just check manuallly the log of the preprod and prod env once this code is in.